### PR TITLE
feat(ci): setup CI to push scorecard, bundle images and run scorecard

### PR DESCRIPTION
**Pull request by tthvo**
# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
* [x] Read the [contributing guidelines](https://github.com/cryostatio/cryostat/blob/main/CONTRIBUTING.md)
* [x] Linked a relevant issue which this PR resolves
* [x] Linked any other relevant issues, PR's, or documentation, if any
* [x] Resolved all conflicts, if any
* [x] Rebased your branch PR on top of the latest upstream `main` branch
* [x] Attached at least one of the following labels to the PR: `[chore, ci, docs, feat, fix, test]`
* [x] [Signed all commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits): `git commit -S -m "YOUR_COMMIT_MESSAGE"`
_______________________________________________

Fixes: #227, #493 
Depends on #589

## Description of the change:

- Add CI to push bundle and scorecard image (if not exists) to quay.io.
- Add CI to run envtest and scorecard.

## Motivation for the change:

See #227, #493

## How to manually test:

See CI job status:
- https://github.com/tthvo/cryostat-operator/actions/runs/5042553727/jobs/9043312859
- https://github.com/tthvo/cryostat-operator/actions/runs/5042639850/jobs/9043480603?pr=4

Push jobs on `main` push is failing on my repo since it does not have access to `cryostat/cryostat` but build steps seems to work fine.

## Others

We also need to remove checks for `build Expected — Waiting for status to be reported ` as CI job name is changed.

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -33,13 +46,112 @@ jobs:
...
+            printf -- "apiVersion: kind.x-k8s.io/v1alpha4\n"
+            printf -- "nodes:\n"
+            printf -- "- role: control-plane\n"
+            printf -- "  kubeadmConfigPatches:\n"
+            printf -- "  - |\n"
+            printf -- "    kind: InitConfiguration\n"
+            printf -- "    nodeRegistration:\n"
+            printf -- "      kubeletExtraArgs:\n"
+            printf -- "        node-labels: \"ingress-ready=true\"\n"
+        } > kind-config.yaml
```
[the remark this answers is not in the document]
Comment (andrewazores) [reply .github/workflows/ci.yaml:109]: I can't find any concrete docs examples but I think this can just be a file at `.github/workflows/kind-config.yaml` for example, and then `kind create cluster` can just reference the file path from the filesystem. I'm not sure what the working directory of the process is so you might need to give it a relative path to the right directory where that file is.

[the remark this answers is not in the document]
Comment (tthvo) [reply .github/workflows/ci.yaml:109]: I think `actions/checkout@v2` should set the current working dir to our repo. Then, it should just be `.github/workflows/kind-config.yaml`.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -33,13 +46,112 @@ jobs:
...
+        kind create cluster --config=kind-config.yaml -n ci-${{ github.repository_owner }}
```
Comment (andrewazores) [inline .github/workflows/ci.yaml:110]: Just for future-proofing in case we have more steps that create such clusters, and in case those clusters aren't created in separate jobs (runner instances), I think the name given to the cluster should be more unique than this.

https://docs.github.com/en/actions/learn-github-actions/contexts#github-context

`github.run_id` or `github.run_number` seem like good candidates.

I think it also makes sense to use the Actions concurrency property:

https://docs.github.com/en/actions/using-jobs/using-concurrency

This is already used in some of our other CI setups to make sure that previous runs get cancelled if a new commit is pushed to the same PR branch, for example. Just to (try to) ensure that we aren't wasting these graciously provided $0 CI resources, but also to keep the CI status uncluttered.

File: internal/test/scorecard/tests.go [hunk 3]
```diff
@@ -288,3 +288,89 @@ func logEvents(r *scapiv1alpha3.TestResult, client *CryostatClientset, namespace
 	}
 	return nil
 }
+
+func newCryostatCR(namespace string, withIngress bool) *operatorv1beta1.Cryostat {
```
Comment (andrewazores) [inline internal/test/scorecard/tests.go:292]: Could you split out these changes to a separate PR? I think they make sense to separate out in the commit history.

File: .github/workflows/ci.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewazores) [reply .github/workflows/ci.yaml:109]: I suppose the config could be useful for other testing with `kind`, including CI run elsewhere than GitHub. A top-level `scripts` or `testing` directory in the repo would make sense to me. @ebaron any opinion?

**Comment by tthvo** [comment]
> I've seen this action before: https://github.com/helm/kind-action. 

Thanks! We still need to pass in the path to config file. Tho, we can fix the version of kind and kubectl. Is there a version we want to set or whatever version (seems like latest) comes with `ubuntu-latest` is fine?

File: internal/test/scorecard/tests.go (hunk 3 above)
Comment (tthvo) [reply internal/test/scorecard/tests.go]: yes just did: https://github.com/cryostatio/cryostat-operator/pull/589

**Comment by tthvo** [comment]
Sounds good! Thanks for that! Then we probably won't use that action as its pretty simple to launch kind.

File: .github/workflows/ci.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ebaron) [reply .github/workflows/ci.yaml:113]: For now I'm leaning towards under `.github/`, Cryostat has similar CI-related things there: https://github.com/cryostatio/cryostat/tree/main/.github

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -33,13 +50,112 @@ jobs:
...
+        kubectl create -n $SCORECARD_NAMESPACE secret docker-registry ghcr-key --docker-server="$GHCR_SERVER" \
+          --docker-username="$GHCR_USERNAME" --docker-password="$GHCR_PASSWORD"
+        kubectl patch sa $SCORECARD_SA_NAME -n $SCORECARD_NAMESPACE -p '{"imagePullSecrets": [{"name": "ghcr-key"}]}'
+
+        # Deploy the operator
+        operator-sdk run bundle -n $SCORECARD_NAMESPACE \
+          --timeout 20m --pull-secret-name ghcr-key --service-account $SCORECARD_SA_NAME ${{ steps.push-bundle-to-ghcr.outputs.registry-path }}
+    - name: Run scorecard tests
+      run: |
+        operator-sdk scorecard -n $SCORECARD_NAMESPACE -s $SCORECARD_SA_NAME -w 20m --pod-security=restricted ./bundle
```
Comment (ebaron) [inline .github/workflows/ci.yaml:149]: I wonder if we modify the `test-scorecard` target to support this and reduce duplication. The differences are optionally setting up the pull secret and then using it when deploying the bundle right?

**Review by ebaron — COMMENTED** [review]
Hey @tthvo, thanks for doing all this. Still working through the steps, but here are some initial comments.

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -33,13 +50,112 @@ jobs:
...
+        } > kind-config.yaml
+        kind create cluster --config=kind-config.yaml -n ci-${{ github.run_id }}
+        # Enabling Ingress
+        kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/main/deploy/static/provider/kind/deploy.yaml
+        kubectl rollout status -w deployment/ingress-nginx-controller -n ingress-nginx --timeout 5m
+    - name: Install Operator Lifecycle Manager
+      run: curl -sL https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.24.0/install.sh | bash -s v0.24.0
+    - name: Install Cert Manager
+      run: make cert_manager
+    - uses: docker/login-action@v2
```
[the remark this answers is not in the document]
Comment (ebaron) [reply .github/workflows/ci.yaml:122]: Does `KIND_EXPERIMENTAL_PROVIDER=podman kind create cluster` work? If not we can stick with Docker.

[the remark this answers is not in the document]
Comment (tthvo) [reply .github/workflows/ci.yaml:122]: ...unfortunately, it does not seem to work:

```
INFO[0000] trying next host                              error="failed to authorize: failed to fetch anonymous token: unexpected status: 401 Unauthorized" host=ghcr.io
FATA[0000] Failed to run bundle: pull bundle image: error pulling image ghcr.io/tthvo/cryostat-operator-bundle:ci-abc78d7db8a6080f9bbe602c01e93e60970f8733: error resolving name for image ref ghcr.io/tthvo/cryostat-operator-bundle:ci-abc78d7db8a6080f9bbe602c01e93e60970f8733: failed to authorize: failed to fetch anonymous token: unexpected status: 401 Unauthorized 
```

[the remark this answers is not in the document]
Comment (tthvo) [reply .github/workflows/ci.yaml:110]: Was testing with my ghcr private image here: `ghcr.io/tthvo/cryostat-operator-bundle:ci-abc78d7db8a6080f9bbe602c01e93e60970f8733`

File: .github/workflows/ci.yaml (hunk 4 above)
Comment (tthvo) [reply .github/workflows/ci.yaml]: I got some difficulty figuring out the clean way for this but I went with: https://github.com/cryostatio/cryostat-operator/pull/588/commits/c0ea55c4ece70b9af5d5202d0189889581054c31.

What do u think?

Comment (ebaron) [reply .github/workflows/ci.yaml]: Nice! This looks pretty much what I had in mind. I have a few more suggestions for it.

File: Makefile [hunk 6]
```diff
@@ -18,7 +18,8 @@ IMAGE_TAG_BASE ?= $(IMAGE_NAMESPACE)/$(OPERATOR_NAME)
 
 # Default bundle image tag
 BUNDLE_IMG ?= $(IMAGE_TAG_BASE)-bundle:$(BUNDLE_VERSION)
-BUNDLE_IMGS ?= $(BUNDLE_IMG) 
+BUNDLE_IMGS ?= $(BUNDLE_IMG)
+BUNDLE_SOURCE ?= $(BUNDLE_IMG)
```
[the remark this answers is not in the document]
Comment (tthvo) [reply Makefile:22]: I don't know if the `operator-sdk scorecard` could pull the bundle image but let me try...

File: Makefile [hunk 7]
```diff
@@ -130,21 +131,33 @@ endif
...
+define scorecard-setup
+@$(CLUSTER_CLIENT) get namespace $(SCORECARD_NAMESPACE) >/dev/null 2>&1 &&\
+	echo "$(SCORECARD_NAMESPACE) namespace already exists, please remove it with \"make clean-scorecard\"" >&2 && exit 1 || true
+$(CLUSTER_CLIENT) create namespace $(SCORECARD_NAMESPACE)
+cd internal/images/custom-scorecard-tests/rbac/ && $(KUSTOMIZE) edit set namespace $(SCORECARD_NAMESPACE)
+$(KUSTOMIZE) build internal/images/custom-scorecard-tests/rbac/ | $(CLUSTER_CLIENT) apply -f -
+
+if [[ -n $(SCORECARD_REGISTRY_SERVER) && -n $(SCORECARD_REGISTRY_USERNAME) && -n $(SCORECARD_REGISTRY_PASSWORD) ]]; then \
+	$(CLUSTER_CLIENT) create -n $(SCORECARD_NAMESPACE) secret docker-registry registry-key --docker-server="$(SCORECARD_REGISTRY_SERVER)" \
+		--docker-username="$(SCORECARD_REGISTRY_USERNAME)" --docker-password="$(SCORECARD_REGISTRY_PASSWORD)"; \
```
Comment (ebaron) [inline Makefile:153]: We should delete this secret in `scorecard-cleanup`.

Comment (tthvo) [reply Makefile:153]: yes thanks!

File: Makefile [hunk 8]
```diff
@@ -130,27 +131,39 @@ endif
...
+	echo "$(SCORECARD_NAMESPACE) namespace already exists, please remove it with \"make clean-scorecard\"" >&2 && exit 1 || true
+$(CLUSTER_CLIENT) create namespace $(SCORECARD_NAMESPACE)
+cd internal/images/custom-scorecard-tests/rbac/ && $(KUSTOMIZE) edit set namespace $(SCORECARD_NAMESPACE)
+$(KUSTOMIZE) build internal/images/custom-scorecard-tests/rbac/ | $(CLUSTER_CLIENT) apply -f -
+
+if [[ -n "$(SCORECARD_REGISTRY_SERVER)" && -n "$(SCORECARD_REGISTRY_USERNAME)" && -n "$(SCORECARD_REGISTRY_PASSWORD)" ]]; then \
+	$(CLUSTER_CLIENT) create -n $(SCORECARD_NAMESPACE) secret docker-registry registry-key --docker-server="$(SCORECARD_REGISTRY_SERVER)" \
+		--docker-username="$(SCORECARD_REGISTRY_USERNAME)" --docker-password="$(SCORECARD_REGISTRY_PASSWORD)"; \
+	$(CLUSTER_CLIENT) patch sa cryostat-scorecard -n $(SCORECARD_NAMESPACE) -p '{"imagePullSecrets": [{"name": "registry-key"}]}'; \
+	$(eval SCORECARD_ARGS=--pull-secret-name registry-key --service-account cryostat-scorecard) \
```
Comment (tthvo) [inline Makefile:155]: Opps this always evaluate. Lets see...

File: Makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (tthvo) [reply Makefile:22]: Seems to be able to pick up the credentials and pull bundle. I will use `ghcr.io` then.

File: .github/workflows/ci.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ebaron) [reply .github/workflows/ci.yaml:108]: It seems like the Operator SDK does indeed hardcode ~/docker/config.json. Can you try with the podman-login action and adding the following parameter: `auth_file_path: ${{ env.HOME }}/.docker/config.json`?

File: Makefile [hunk 9]
```diff
@@ -145,12 +140,29 @@ endif
...
+define scorecard-setup
+@$(CLUSTER_CLIENT) get namespace $(SCORECARD_NAMESPACE) >/dev/null 2>&1 &&\
+	echo "$(SCORECARD_NAMESPACE) namespace already exists, please remove it with \"make clean-scorecard\"" >&2 && exit 1 || true
+$(CLUSTER_CLIENT) create namespace $(SCORECARD_NAMESPACE)
+cd internal/images/custom-scorecard-tests/rbac/ && $(KUSTOMIZE) edit set namespace $(SCORECARD_NAMESPACE)
+$(KUSTOMIZE) build internal/images/custom-scorecard-tests/rbac/ | $(CLUSTER_CLIENT) apply -f -
+
+if [[ -n "$(SCORECARD_REGISTRY_SERVER)" && -n "$(SCORECARD_REGISTRY_USERNAME)" && -n "$(SCORECARD_REGISTRY_PASSWORD)" ]]; then \
+	$(CLUSTER_CLIENT) create -n $(SCORECARD_NAMESPACE) secret docker-registry registry-key --docker-server="$(SCORECARD_REGISTRY_SERVER)" \
+		--docker-username="$(SCORECARD_REGISTRY_USERNAME)" --docker-password="$(SCORECARD_REGISTRY_PASSWORD)"; \
```
Comment (ebaron) [inline Makefile:152]: I noticed that the password is logged here. Maybe prefix this with `@`

File: Makefile (hunk 8 above)
Comment (ebaron) [reply Makefile:154]: This would work without the `eval` wouldn't it? Then you just need to add an extra `$` when you use it to escape Make

Comment (tthvo) [reply Makefile:154]: Then the `operator-sdk run bundle` should run in the same shell as the if-else right?

Comment (ebaron) [reply Makefile:154]: Ah, I thought it was already. What's the reasoning behind using a Bash if-statement instead of Make btw?

Comment (tthvo) [reply Makefile:154]: ah i just had not found a way to do multiple checks with Makefile. I tried `and` function but its not behaving like I thought... Any suggestion?

Comment (ebaron) [reply Makefile:154]: Seems like the way to do it is 3 nested `ifneq` `endif` blocks. Not the prettiest, but what can you do. Then we could have each command execute in its own shell and make `SCORECARD_ARGS` a Make variable right?

Comment (tthvo) [reply Makefile:154]: yeh seems like it, I will update it...

Comment (tthvo) [reply Makefile:154]: opps nvmind I think we can use `and` function to check. There was a typo before.

```
ifneq ($(and $(SCORECARD_REGISTRY_SERVER),$(SCORECARD_REGISTRY_USERNAME),$(SCORECARD_REGISTRY_PASSWORD)),)
SCORECARD_ARGS := --pull-secret-name registry-key --service-account cryostat-scorecard
endif
```

File: .github/workflows/ci.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (tthvo) [reply .github/workflows/ci.yaml]: I am running into this issue on CI:

```
Run redhat-actions/podman-login@v1
/usr/bin/podman version
  /usr/bin/podman version
  Version:      3.4.4
  API Version:  3.4.4
  Go Version:   go1.17.3
  Built:        Thu Jan  1 00:00:00 1970
  OS/Arch:      linux/amd64
/usr/bin/podman login ghcr.io -u tthvo -p *** --verbose --authfile=/.docker/config.json
Error: 1 error occurred:
	* mkdir /.docker: permission denied


Error: Error: podman exited with code 125
Error: 1 error occurred:
	* mkdir /.docker: permission denied
```

[the remark this answers is not in the document]
Comment (andrewazores) [reply .github/workflows/ci.yaml]: `env.HOME` is undefined/empty, I guess?

https://docs.github.com/en/actions/learn-github-actions/variables

[the remark this answers is not in the document]
Comment (andrewazores) [reply .github/workflows/ci.yaml]: `RUNNER_TEMP` might be useful?

[the remark this answers is not in the document]
Comment (tthvo) [reply .github/workflows/ci.yaml]: Oh wow right! Lets try that!

[the remark this answers is not in the document]
Comment (andrewazores) [reply .github/workflows/ci.yaml]: Weird.

image

[the remark this answers is not in the document]
Comment (andrewazores) [reply .github/workflows/ci.yaml]: Maybe the `${{ env.FOO }}` syntax is only for env vars that are defined on the org/repo?

image

Perhaps just try using `$RUNNER_TEMP` instead of `${{ env.RUNNER_TEMP }}`

[the remark this answers is not in the document]
Comment (andrewazores) [reply .github/workflows/ci.yaml]: I guess that's what this section is talking about, but I haven't really read it:

https://docs.github.com/en/actions/learn-github-actions/variables#using-the-env-context-to-access-environment-variable-values

[the remark this answers is not in the document]
Comment (ebaron) [reply .github/workflows/ci.yaml]: Or maybe just `$HOME`?

[the remark this answers is not in the document]
Comment (tthvo) [reply .github/workflows/ci.yaml]: Yess `$HOME` worked!

[the remark this answers is not in the document]
Comment (ebaron) [reply .github/workflows/ci.yaml]: Awesome! Thanks for sticking with this. I prefer using Podman/Buildah tooling where possible.

File: Makefile [hunk 10]
```diff
@@ -145,12 +140,31 @@ endif
...
+SCORECARD_ARGS := --pull-secret-name registry-key --service-account cryostat-scorecard
+endif
+
+define scorecard-setup
+@$(CLUSTER_CLIENT) get namespace $(SCORECARD_NAMESPACE) >/dev/null 2>&1 &&\
+	echo "$(SCORECARD_NAMESPACE) namespace already exists, please remove it with \"make clean-scorecard\"" >&2 && exit 1 || true
+$(CLUSTER_CLIENT) create namespace $(SCORECARD_NAMESPACE)
+cd internal/images/custom-scorecard-tests/rbac/ && $(KUSTOMIZE) edit set namespace $(SCORECARD_NAMESPACE)
+$(KUSTOMIZE) build internal/images/custom-scorecard-tests/rbac/ | $(CLUSTER_CLIENT) apply -f -
+if [ -n "$(SCORECARD_ARGS)" ]; then \
```
Comment (ebaron) [inline Makefile]: I'm getting a Make error here. Are we able to use `ifneq` here instead of Bash?

Comment (tthvo) [reply Makefile]: oh yeh I just updated that to silent the entire if. I tried `ifneq` but got syntax error all over...

Comment (tthvo) [reply Makefile]: Worked here: https://github.com/tthvo/cryostat-operator/actions/runs/5074021438/jobs/9113746052

Just need that k8s support for scorecard :))

**Review by ebaron — COMMENTED** [review]
One thing I noticed that's missing here is a CI image for the operator controller image. This means that any code changes aren't reflected in the bundle image, only YAML changes. We should push the operator image to ghcr.io and then pass that image to `make bundle` with `OPERATOR_IMG`.

**Comment by tthvo** [comment]
> We should push the operator image to ghcr.io and then pass that image to make bundle with OPERATOR_IMG.

Do u think we should patch the manager's pod template or the operator SA with the pullImageSecret?

**Comment by tthvo** [comment]
Not the most intuitive way to patch imagePullSecrets but seems to work nicely:

Patch: https://github.com/cryostatio/cryostat-operator/pull/588/commits/46e6741382ecf68afcbeab0e7f4750a0025576db

Test run:  https://github.com/tthvo/cryostat-operator/actions/runs/5083194894/jobs/9133896077

How about that?

**Comment by tthvo** [comment]
@ebaron Latest run after rebasing: https://github.com/tthvo/cryostat-operator/actions/runs/5083932753/jobs/9135558209

**Review by ebaron — APPROVED** [review]
Works well! Awesome job on getting this done @tthvo! Thank you!

**Comment by andrewazores** [comment]
This is awesome, finally knocking off a major piece of that puzzle for doing CI in a k8s environment.

**Comment by tthvo** [comment]
Niceee thanks you^^ Happy I could help with it :))
